[manuf] Add test for the manuf_cp_test_lock testpoint

jwnrt opened the pull request:
This PR adds a manufacturing test for the `manuf_cp_test_lock` testpoint. Resolves https://github.com/lowRISC/opentitan/issues/17390.

It includes the OTP_CTRL test utility commits from https://github.com/lowRISC/opentitan/pull/18471 which should be merged first separately.

The final two commits are the new ones:

- Fixes LC controller errors not being reported properly.
- Adds manufacturing test for `manuf_cp_test_lock`:
  - Starts with device in `DEV` lifecycle state.
  - Generates a `TEST_UNLOCK` token pre-image and post-image with `gen_test_unlock_token.py`.
  - Programs the token post-image to OTP.
  - Reads back the token to verify the program.
  - Locks the SECRET0 OTP partition and resets.
  - Transitions to `TEST_LOCKED0`.
  - Separately, transitions to `TEST_UNLOCKED1` using the token pre-image.

The test gets most of the way, but unlocking gives a `TOKEN_ERROR` from the lifecycle controller.

Still trying to work out why this is.

timothytrippel commented on sw/device/silicon_creator/manuf/lib/BUILD:
```diff
@@ -483,3 +483,23 @@ opentitan_functest(
+    name = "manuf_cp_test_lock_functest",
+    srcs = ["//sw/device/silicon_creator/rom/e2e:empty_test.c"],
+    cw310 = cw310_params(
+        bitstream = "//hw/bitstream:rom_otp_test_unlocked0",
```
Use the OTP image I describe here: https://github.com/lowRISC/opentitan/pull/18471#discussion_r1191529819 instead as it will be completely empty, with ROM execution disabled. Just add it to #18471 and rebase this PR on top to use it too.

jwnrt replied:
Thanks, this with the bootstrapping removed fixed the errors I was having.

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,155 @@
+        "check correct TEST_UNLOCK token can be read back from OTP"
+    );
+
+    jtag.disconnect()?;
```
to be explicit, you could remove the CPU TAP straps here after disconnecting

timothytrippel replied:
same throughout

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,155 @@
+    lc_transition::trigger_lc_transition(
+        transport,
+        jtag.clone(),
+        DifLcCtrlState::TestUnlocked0,
```
If you start in TEST_UNLOCKED0, [you can only transition to TEST_LOCKED0](https://opentitan.org/book/doc/security/specs/device_life_cycle/index.html#architecture). I think you want to transition to TEST_UNLOCKED1 here.

jwnrt replied:
Ah, this explains all those LC states. Thanks.

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,155 @@
+        true,
+        opts.init.bootstrap.options.reset_delay,
+    )
+    .context("failed to trigger transition to TEST_UNLOCKED0")?;
```
nit: should be "TEST_LOCKED0"

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,155 @@
+    // Read back the token to verify it was written.
+    let mut test_unlock_token = [0xff; 4];
+    OtpParam::read_param(&*jtag, DamParam::TestUnlockToken, &mut test_unlock_token)
+        .context("failed to read back TEST_UNLOCK token from OTP")?;
```
I believe you need to lock this partition, and trigger a reset, before the token will become active. Though I can't seem to locate this in the docs.

@msfschaffner : for the secret0 partition, do you need to lock it / perform a reset to complete the locking after provisioning the TEST_UNLOCK/EXIT tokens for them to be ungated to the LC ctrl? Can't seem to find this mentioned in the docs.

jwnrt replied:
I've updated the `provision_token` step to lock `SECRET0` and reset after provisioning, but the LC controller is still giving a status with `TOKEN_ERROR` set when trying to unlock.

Do I need to make `TEST_EXIT` non-zero before locking or anything like that?

timothytrippel replied:
In the `manuf_cp_device_flash_info_wr` test I set the test exit and unlock tokens to the same value, so you could just do that

timothytrippel reviewed (commented):
I only reviewed the last commit, I agree lets get #18471 merged first.

Couple of suggestions otherwise mostly LGTM, thanks @jwnrt !

jwnrt commented:
Thanks for the review @timothytrippel 

With the OTP changes, this gets most of the way through the test now.

Still getting a `TOKEN_ERROR` bit set when doing the unlock transition, so something must be wrong with my token. I'll double check that.

timothytrippel commented:
awesome, perhaps it is a token endianness issue? not sure if you checked the by ordering?

timothytrippel commented on sw/device/silicon_creator/manuf/lib/BUILD:
```diff
@@ -518,3 +518,23 @@ opentitan_functest(
+    ),
+    data = OPENTITANTOOL_OPENOCD_DATA_DEPS,
+    targets = ["cw310_rom"],
+    test_harness = "//sw/host/tests/manuf/manuf_cp_test_lock:manuf_cp_test_lock",
```
nit: some Bazel syntactic sugar, if the target name matches the package name in the label you can drop the former
```suggestion
    test_harness = "//sw/host/tests/manuf/manuf_cp_test_lock"
```

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,171 @@
+
+//! This test performs a lifecycle transition from `TEST_UNLOCKED0` to `TEST_LOCKED0`.
+//!
+//! We also test provisioning a `TEST_UNLOCKED` being programmed into OTP.
```
What do you mean by this?

jwnrt replied:
I just meant that programming the chip to OTP (provisioning) is done within the test via JTAG rather than preloaded into the OTP before the test starts.

Updated the wording so hopefully it makes sense now

timothytrippel replied:
ah got it, thanks.

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,171 @@
+/// Post-image of the TEST_UNLOCK token as generated by `gen_test_unlock_token.py`.
+const TEST_UNLOCK_TOKEN_POSTIMAGE: [u32; 4] = [0xe6b9ec3f, 0x67fd75f3, 0x618ae065, 0xc0bb8f81];
+
+fn main() -> anyhow::Result<()> {
```
nit: import "anyhow::Result"?

jwnrt replied:
Sure. I only write these out to avoid confusion with the standard `Result` from the prelude, but don't mind importing it.

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,171 @@
+    assert_ne!(digest, [0x00, 2]);
+
+    jtag.disconnect()?;
+    transport.pin_strapping("PINMUX_TAP_RISCV")?.apply()?;
```
don't you want:
```suggestion
transport.pin_strapping("PINMUX_TAP_RISCV")?.remove()?;
```

?

jwnrt replied:
Oops yes I do

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,171 @@
+    assert_eq!(state, DifLcCtrlState::TestLocked0.redundant_encoding());
+
+    jtag.disconnect()?;
+    transport.pin_strapping("PINMUX_TAP_LC")?.apply()?;
```
same here `.remove()` not `.apply`?

timothytrippel commented on sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs:
```diff
@@ -0,0 +1,171 @@
+    assert_eq!(state, DifLcCtrlState::TestUnlocked1.redundant_encoding());
+
+    jtag.disconnect()?;
+    transport.pin_strapping("PINMUX_TAP_LC")?.apply()?;
```
same here remove not apply

timothytrippel reviewed (approved):
LGTM, now that #18471 is merged can you rebase this? then I can pull it down to see if I can figure out what is going on.

jwnrt commented:
Rebased on the new OTP test util and fixed up review comments.

I tried setting the `TEST_EXIT` token before locking, too, but we get the same error back from the lc_ctrl

<details>
<summary>patch (show/hide)</summary>

```patch
diff --git a/sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs b/sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs
index 70225fa93..ad541c0e3 100644
--- a/sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs
+++ b/sw/host/tests/manuf/manuf_cp_test_lock/src/main.rs
@@ -58,6 +58,13 @@ fn provision_token(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
     )
     .context("failed to write TEST_UNLOCK token to OTP")?;
 
+    OtpParam::write_param(
+        &*jtag,
+        DaiParam::TestExitToken,
+        &TEST_UNLOCK_TOKEN_POSTIMAGE,
+    )
+    .context("failed to write TEST_EXIT token to OTP")?;
+
     // Read back the token to verify it was written.
     let mut test_unlock_token = [0xff; 4];
     OtpParam::read_param(&*jtag, DaiParam::TestUnlockToken, &mut test_unlock_token)
@@ -68,6 +75,16 @@ fn provision_token(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
         "check correct TEST_UNLOCK token can be read back from OTP"
     );
 
+    // Read back the token to verify it was written.
+    let mut test_exit_token = [0xff; 4];
+    OtpParam::read_param(&*jtag, DaiParam::TestExitToken, &mut test_exit_token)
+        .context("failed to read back TEST_EXIT token from OTP")?;
+
+    assert_eq!(
+        test_exit_token, TEST_UNLOCK_TOKEN_POSTIMAGE,
+        "check correct TEST_EXIT token can be read back from OTP"
+    );
+
     // Lock the SECRET0 partition.
     OtpPartition::lock(&*jtag, Partition::SECRET0).context("failed to lock SECRET0 partition")?;
 ```

</details>

timothytrippel commented:
Awesome thanks @jwnrt , I will pull this down and see if I can uncover the issue here.

jwnrt commented:
> awesome, perhaps it is a token endianness issue? not sure if you checked the by ordering?

I split the postimage that `gen_test_unlock_token.py` gives into words directly, i.e.

`0xe6b9ec3f67fd75f3618ae065c0bb8f81 -> [0xe6b9ec3f, 0x67fd75f3, 0x618ae065, 0xc0bb8f81]`.

Is that okay? Do I need to change the word order or byte order? I'm not able to test on an FPGA right now

timothytrippel commented:
Yep, I pulled down this test and confirmed, you just have the token words in the wrong order, I reversed the order and the test passes now. I can repush it with the fix.

timothytrippel commented:
Just repushed the update, I can merge once it passes CI.
